app/controllers/passwords_controller.rb

```diff
@@ -0,0 +1,41 @@
+class PasswordsController < ApplicationController
+
+  def new
+  end
+
+  def create
+    email = params[:email]
+    user = User.find_by_email(email)
+    if user
+      user.reset_password_token = SecureRandom.urlsafe_base64
+      user.save
+      AppMailer.send_mail_on_reset_password(user).deliver
+      render 'passwords/confirm'
+    end
+  end
+
+  def edit
+    @token = params[:token]
+    @user = User.find_by_reset_password_token(@token)
+    if !@user
+      flash[:error] = 'Your reset password link is expired.'
+    end
+  end
+
+  def update
+    token = params[:token]
+    password = params[:password]
+    user = User.find_by_reset_password_token(token)
+    if !user
```

Comment:
Your happy path should be first. I would remove the bang and swap the order of the clauses.